### ISSUE TITLE
Losing connection whenever the user is using my app while playing a song.

### ISSUE THREAD

**Issue by Slin** [spotify/ios-sdk#45]
I can play songs from within my app and if I do nothing after hitting play it works ok, but whenever I scroll around to look at other songs I get the following error and playback stops:

`AppRemote: Disconnected with error: Error Domain=com.spotify.app-remote.transport Code=-2000 "Stream error." UserInfo={NSLocalizedDescription=Stream error., NSUnderlyingError=0x283d6e220 {Error Domain=NSPOSIXErrorDomain Code=32 "Broken pipe" UserInfo={_kCFStreamErrorCodeKey=32, _kCFStreamErrorDomainKey=1}}, NSLocalizedRecoverySuggestion=Reconnect the transport to the Spotify app., NSLocalizedFailureReason=A stream error has occured in the transport.}`

Memory and CPU utilization is a little higher at those moments then otherwise, but not exactly high either (130mb, 60% cpu utilization). This worked ok 3 weeks ago. I did switch from an iPhone 7+ to an iPhone Xs in that same time though.

**Comment by yuriyhanysh** [comment]
@jackfreeman I get the same error, although I'm not scrolling around, simply starting the test application.
This error is not 100% reproducible, after a few hours it actually managed to connect to the Spotify app and I didn't change anything in the code.

My environment:
iPhone 6s with iOS 12.0.
Spotify app is the latest from the AppStore. Music is playing.
Spotify SDK v1.0.0
As a test app, I'm using NowPlayingView from SDK DemoProjects folder.
There are no major changes in it, I've replaced ClientID with id from my dashboard and added LSApplicationQueriesSchemes key into Info.plist.

[SpotifySDKConnectionError.zip](https://github.com/spotify/ios-sdk/files/2510677/SpotifySDKConnectionError.zip)
In the video, you can see that when this happens, Waze cannot control Spotify as well (but I don't know if Waze uses the same SDK).
While I can reproduce this error is there somehow I can help to investigate this bug (systems logs etc.)?

Also, here are logs from NowPlayingView app (with debug log level):

```
2018-10-24 15:38:00.202443+0300 NowPlayingView[3124:900811] AppRemote: Connecting...
2018-10-24 15:38:00.230590+0300 NowPlayingView[3124:900876] [] nw_socket_handle_socket_event [C1.1:1] Socket SO_ERROR [61: Connection refused]
2018-10-24 15:38:00.234347+0300 NowPlayingView[3124:900876] [] nw_socket_handle_socket_event [C1.2:1] Socket SO_ERROR [61: Connection refused]
2018-10-24 15:38:00.234633+0300 NowPlayingView[3124:900875] [] nw_connection_get_connected_socket [C1] Client called nw_connection_get_connected_socket on unconnected nw_connection
2018-10-24 15:38:00.235124+0300 NowPlayingView[3124:900875] TCP Conn 0x2807b3840 Failed : error 0:61 [61]
2018-10-24 15:38:00.235761+0300 NowPlayingView[3124:900811] AppRemote: Failed connection attempt with error: Error Domain=com.spotify.app-remote.transport Code=-2000 "Stream error." UserInfo={NSLocalizedDescription=Stream error., NSUnderlyingError=0x283ce49c0 {Error Domain=NSPOSIXErrorDomain Code=61 "Connection refused" UserInfo={_kCFStreamErrorCodeKey=61, _kCFStreamErrorDomainKey=1}}, NSLocalizedRecoverySuggestion=Reconnect the transport to the Spotify app., NSLocalizedFailureReason=A stream error has occured in the transport.}
didFailConnectionAttemptWithError Optional(Error Domain=com.spotify.app-remote Code=-1000 "Connection attempt failed." UserInfo={NSLocalizedDescription=Connection attempt failed., NSUnderlyingError=0x283ce48d0 {Error Domain=com.spotify.app-remote.transport Code=-2000 "Stream error." UserInfo={NSLocalizedDescription=Stream error., NSUnderlyingError=0x283ce49c0 {Error Domain=NSPOSIXErrorDomain Code=61 "Connection refused" UserInfo={_kCFStreamErrorCodeKey=61, _kCFStreamErrorDomainKey=1}}, NSLocalizedRecoverySuggestion=Reconnect the transport to the Spotify app., NSLocalizedFailureReason=A stream error has occured in the transport.}}, NSLocalizedRecoverySuggestion=Ensure the Spotify app is installed, and try to reconnect., NSLocalizedFailureReason=Could not connect to the Spotify app.})
2018-10-24 15:38:10.393849+0300 NowPlayingView[3124:900811] AppRemote: Disconnecting...
2018-10-24 15:38:11.257800+0300 NowPlayingView[3124:900811] Could not load IOSurface for time string. Rendering locally instead.
2018-10-24 15:38:11.327951+0300 NowPlayingView[3124:900811] Could not load IOSurface for time string. Rendering locally instead.
2018-10-24 15:38:12.854556+0300 NowPlayingView[3124:900811] AppRemote: Connecting...
2018-10-24 15:38:12.874209+0300 NowPlayingView[3124:900910] [] nw_socket_handle_socket_event [C2.1:1] Socket SO_ERROR [61: Connection refused]
2018-10-24 15:38:12.876818+0300 NowPlayingView[3124:900910] [] nw_socket_handle_socket_event [C2.2:1] Socket SO_ERROR [61: Connection refused]
2018-10-24 15:38:12.877184+0300 NowPlayingView[3124:900883] [] nw_connection_get_connected_socket [C2] Client called nw_connection_get_connected_socket on unconnected nw_connection
2018-10-24 15:38:12.877204+0300 NowPlayingView[3124:900883] TCP Conn 0x2807b4840 Failed : error 0:61 [61]
2018-10-24 15:38:12.877552+0300 NowPlayingView[3124:900811] AppRemote: Failed connection attempt with error: Error Domain=com.spotify.app-remote.transport Code=-2000 "Stream error." UserInfo={NSLocalizedDescription=Stream error., NSUnderlyingError=0x283ced470 {Error Domain=NSPOSIXErrorDomain Code=61 "Connection refused" UserInfo={_kCFStreamErrorCodeKey=61, _kCFStreamErrorDomainKey=1}}, NSLocalizedRecoverySuggestion=Reconnect the transport to the Spotify app., NSLocalizedFailureReason=A stream error has occured in the transport.}
didFailConnectionAttemptWithError Optional(Error Domain=com.spotify.app-remote Code=-1000 "Connection attempt failed." UserInfo={NSLocalizedDescription=Connection attempt failed., NSUnderlyingError=0x283ced8c0 {Error Domain=com.spotify.app-remote.transport Code=-2000 "Stream error." UserInfo={NSLocalizedDescription=Stream error., NSUnderlyingError=0x283ced470 {Error Domain=NSPOSIXErrorDomain Code=61 "Connection refused" UserInfo={_kCFStreamErrorCodeKey=61, _kCFStreamErrorDomainKey=1}}, NSLocalizedRecoverySuggestion=Reconnect the transport to the Spotify app., NSLocalizedFailureReason=A stream error has occured in the transport.}}, NSLocalizedRecoverySuggestion=Ensure the Spotify app is installed, and try to reconnect., NSLocalizedFailureReason=Could not connect to the Spotify app.})
2018-10-24 15:38:17.530368+0300 NowPlayingView[3124:900811] AppRemote: Disconnecting...
2018-10-24 15:38:18.136797+0300 NowPlayingView[3124:900811] Could not load IOSurface for time string. Rendering locally instead.
2018-10-24 15:38:18.215584+0300 NowPlayingView[3124:900811] Could not load IOSurface for time string. Rendering locally instead.
```

**Comment by Slin** [comment]
I'm still investigating the issue on my side because it seems like not doing anything Spotify related while music is playing, somewhat solves that issue. It seems like the sdk is just directly talking to the Spotify app whenever a call is made and it's not queuing anything and taking care of the order of requests or anything and the Spotify app doesn't seem to like that very much.

I had it perfectly working on my old phone, than I got a new one and now I keep losing connection again, which I think is timing issue or the result of too many requests or something...
I also have hours when everything seems to work fine and then it suddenly just stops.

My current plan is write a wrapper around the sdk, which takes care of only doing one request at a time and will only do the next one once the previous finished.

**Comment by jackfreeman** [comment]
@yuriyhanysh Do you have any steps to reproduce the issue?

**Comment by yuriyhanysh** [comment]
@jackfreeman unfortunately no. I'll try to find some but, as I wrote above, this problem was pretty random. Right now everything seems OK.

**Comment by j-hit** [comment]
Related to https://github.com/spotify/ios-sdk/issues/55

**Comment by Slin** [comment]
Turns out my issue was the AVAudioSession, which I accidentally set to be app only whenever I tried playing something and then there were some videos playing inside the app (without audio) which ended up interrupting the Spotify apps playback. This was still remaining (and quite hidden behind some abstractions) from using the old SDK before. Setting the audio session to mixed playback everything now started working. Still dealing with some issues, but I should be able to solve those on my side.
